### PR TITLE
[RFC] Poor Man's Cron

### PR DESCRIPTION
**Pull request by alikon**
porting of https://github.com/joomla-projects/privacy-framework/pull/133
Pull Request for Issue #20839

### Summary of Changes 
a basic job scheduler system  to be able to schedule and execute tasks that could be run as:

- a cron job on CLI/terminal 
- a webcron from webcron services
- a normal `onAfterRespond` event

plugins.

- a new  system plugin `scheduler` 

it runs at scheduled time
it trigger the  `onExecuteScheduledTask` event from the new job plugin group
can be triggered from a webcron service

- a new  job  plugin `logrotation`

when triggered (even from cli/cron) it runs at scheduled time

the CLI script

- a new cli  `scheduler` script

it trigger the  `onExecuteScheduledTask` event from the new job plugin group

### Testing Instructions

apply PR and discover 
install the system scheduler plugin
install the job logrotation plugin
disable the system log rotation plugin
enable the debug

additionally you can install these 2 sample job plugin
[plg_jobone.zip](https://github.com/joomla/joomla-cms/files/2430424/plg_jobone.zip)
[plg_jobtwo.zip](https://github.com/joomla/joomla-cms/files/2430426/plg_jobtwo.zip)

wich do quite nothing, just sleep for some seconds, useful to help testing concurrency

#### Test settings

**`the scheduler system plugin`**

screenshot from 2018-09-29 10-51-03

**`the job plugin`**

screenshot from 2018-09-29 11-00-58

i can suggest to test with the  _**ApacheBench**_ tool

if you want to test the webcron way
`ab -n 1000 -c 50 http://yoursite/?webcronkey=mysecretkey`
or without a webcron
`ab -n 1000 -c 50 http://yoursite/`

for those lucky enough to use a cron job simply schedule/run from CLI
`php cli/scheduler.php`

the execution can be checked in the log path 
and should look like something like this , depending on your plugin settings
screenshot from 2018-09-29 11-05-31

### Expected result
even without a cron facilities you are able to schedule job's

### Actual result
n.a

### Additional comments

it can be easly used for tasks like:

- deleteExpiredConsents
- remindExpiringConsents
- logRotation
- updateNotifications
- sessionGC

etc...

### todo
if approved
 convert the old plugins that use directly `onAfterRespond` event
 the sql install/update script

### CODE REVIEW

File: plugins/job/logrotation/logrotation.xml [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+				<field
+					name="logstokeep"
+					type="integer"
+					label="PLG_JOB_LOGROTATION_LOGSTOKEEP_LABEL"
+					description="PLG_JOB_LOGROTATION_LOGSTOKEEP_DESC"
+					first="1"
+					last="10"
+					step="1"
+					default="1"
+					filter="int"
```
Comment (Quy) [inline plugins/job/logrotation/logrotation.xml]: change to `integer`

File: plugins/job/logrotation/logrotation.xml [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+				<field
+					name="cachetimeout"
+					type="integer"
+					label="PLG_JOB_LOGROTATION_CACHETIMEOUT_LABEL"
+					description="PLG_JOB_LOGROTATION_CACHETIMEOUT_DESC"
+					first="0"
+					last="120"
+					step="1"
+					default="30"
+					filter="int"
```
Comment (Quy) [inline plugins/job/logrotation/logrotation.xml]: change to `integer`

File: plugins/job/logrotation/logrotation.xml [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+					validate="number"
+				/>
+
+				<field
+					name="lastrun"
+					type="hidden"
+					default="0"
+					filter="integer"
+				/>
+				<field
```
Comment (Quy) [inline plugins/job/logrotation/logrotation.xml:55]: Add blank line above

File: plugins/system/scheduler/scheduler.xml [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+				<field
+					name="cachetimeout"
+					type="integer"
+					label="PLG_SYSTEM_SCHEDULER_CACHETIMEOUT_LABEL"
+					description="PLG_SYSTEM_SCHEDULER_CACHETIMEOUT_DESC"
+					first="0"
+					last="5"
+					step="1"
+					default="3"
+					filter="int"
```
Comment (Quy) [inline plugins/system/scheduler/scheduler.xml:31]: change to `integer`

File: plugins/system/scheduler/scheduler.xml [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+					validate="number"
+				/>
+				<field
+					name="lastrun"
+					type="hidden"
+					default="0"
+					size="15"
```
Comment (Quy) [inline plugins/system/scheduler/scheduler.xml:62]: This is `hidden` type, no need to specify `size`.

File: plugins/job/logrotation/logrotation.xml [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+		<language tag="en-GB">en-GB.plg_job_logrotation.ini</language>
+		<language tag="en-GB">en-GB.plg_job_logrotation.sys.ini</language>
+	</languages>
+	<config>
+		<fields name="params">
+			<fieldset name="basic">
+				<field
+					name="cachetimeout"
+					type="integer"
+					label="PLG_JOB_LOGROTATION_CACHETIMEOUT_LABEL"
```
Comment (Quy) [inline plugins/job/logrotation/logrotation.xml]: Where is the language file for these strings?

File: cli/scheduler.php [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+require_once JPATH_CONFIGURATION . '/configuration.php';
+
+// System configuration.
+$config = new JConfig;
+define('JDEBUG', $config->debug);
+
+// Configure error reporting to maximum for CLI output.
+error_reporting(E_ALL);
+ini_set('display_errors', 1);
+use Joomla\Registry\Registry;
```
Comment (Quy) [inline cli/scheduler.php]: Add blank line above the use statement.

**Comment by alikon** [comment]
@Quy i will be happy to fix all CS related issues, 
just need some kind of green light before :roll_eyes:

File: plugins/job/logrotation/logrotation.php [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+	 * The log check and rotation code event.
+	 *
+	 * @return  void
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function onExecuteScheduledTask($task = array())
+	{
+
+
```
Comment (csthomas) [inline plugins/job/logrotation/logrotation.php:70]: IMO the lock method should be before executing job.

I prefer something like this
```
        /* If success then job can start */
	protected function lockPlugin($id, $checked_out='1') {
		$db = JFactory::getDbo();
		$checked_out_time = date_create('@'.time())->format('Y-m-d H:i:s');
		$minSpace = date_create('@'.(time()-180))->format('Y-m-d H:i:s'); // 3m
		$q = $db->getQuery(true)
			->update($db->qn('#__extensions'))
			->set($db->qn('checked_out').' = '.$db->q($checked_out))
			->set($db->qn('checked_out_time').' = '.$db->q($checked_out_time))
			->where($db->qn('extension_id').' = '.$db->q($id))
			->where('('.$db->qn('checked_out').' = 0'.
			        ' OR ('.$db->qn('checked_out').' = '.$db->q($checked_out).
			        ' AND '.$db->qn('checked_out_time').' < '.$db->q($minSpace).'))');
		$db->setQuery($q);
		$db->execute();
		return $db->getAffectedRows();
	}

	protected function unlockPlugin($id, $checked_out='1') {
		$db = JFactory::getDbo();
		$q = $db->getQuery(true)
			->update($db->qn('#__extensions'))
			->set($db->qn('checked_out').' = '.$db->q('0'))
			->set($db->qn('checked_out_time').' = '.$db->q('0000-00-00 00:00:00'))
			->where($db->qn('extension_id').' = '.$db->q($id))
			->where($db->qn('checked_out').' IN (0,'.$db->q($checked_out).')');
		$db->setQuery($q);
		return $db->execute();
	}
```

**Comment by roland-d** [comment]
@alikon Personally I am not so in favor for the poor man's cron but I am not against it though. Let me explain. First of all, if a host doesn't offer cron job facilities, it's time to look for another host. Second, it doesn't offer the same options as a real cron job. You cannot run the Joomla CLI files as we check if you are on the CLI. You cannot be sure the job is executed at the set time, you are depending on the visitor. Third, the user who happens to visit the site and trigger the job has to wait until the job is finished. Fourth, you are limited to the max_execution_time of your host. If your host doesn't offer real cron jobs, this is most likely not exceeding the 30 second mark.

There is also a service for this like https://www.easycron.com/ which can trigger cronjobs if your host doesn't support it.

Users usually don't understand these limitations and if this goes in we should be very clear about this. As an extension developer that relies a lot on cronjobs I have seen the problems that these poor man crons give. In my case it is users importing huge amounts of data and wondering why it crashes or takes so long. I can see the questions coming that the cron doesn't work.

Just my 3 cents :)

**Comment by brianteeman** [comment]
@roland-d I agree with all your points. No question about it - a real cron will always be better. However this is much more beginner friendly and for those people who regularly move hosts it has the advantage of being portable. As long as there is a notice pointing out that a real cron is better etc then I don't see an issue with this.

**Comment by brianteeman** [comment]
@roland-d should be noticed that drupal has a "poor mans cron" see https://www.drupal.org/docs/8/cron-automated-tasks/cron-automated-tasks-overview

**Comment by mbabker** [comment]
There is also a similar capability in WordPress - https://developer.wordpress.org/plugins/cron/

**Comment by mbabker** [comment]
> As an extension developer that relies a lot on cronjobs I have seen the problems that these poor man crons give.

Similar to the session GC capabilities added in 3.8, when complete this should be able to run in a way where the actual task runners can execute in either a web context or through a CLI script (then someone who can and does set up a proper cron at the server level can turn off the web based executor, probably a plugin, and rely on the server's cron to run things).

**Comment by alikon** [comment]
sorry folks for my late reply.... even if i'm lazy.... i'm still  working on this task (with my times, sorry ) ... for me this it is a **very important task**
@roland-d all your concerns are mine's too... and the same i think for the other's who cares about this matter

for sure  we cannot equiparate a cron with a web-cron  we all agree on this...

and this must be crystal clear when and if this [wip] pr will go in to the core

...my simulations with tools like _ApacheBench_ are smoothly going quite well, my first goal is to manage concurrency for the web side... for the cron side if we a have a normal sysadmin ... maybe we should'nt need to manage that issue.... 

i'll come back with some still "new  and dirty code" quite soon...

**Comment by alikon** [comment]
updated the pr description&test info accordingly to the latest changes

File: libraries/src/Plugin/PluginHelper.php [hunk 9]
```diff
@@ -365,4 +367,176 @@ protected static function load()
...
+			->update($db->quoteName('#__extensions'))
+			->set($db->qn('checked_out_time') . ' = ' . $db->q('1918-01-01 00:00:00'))
+			->where($db->quoteName('element') . ' = ' . $db->quote($name))
+			->where($db->quoteName('folder') . ' = ' . $db->quote($type))
+			->where($db->quoteName('checked_out_time') . ' != ' . $db->q('1918-01-01 00:00:00'));
+
+		try
+		{
+			// Lock the tables to prevent multiple plugin executions causing a race condition
+			$db->lockTable('#__extensions');
```
Comment (csthomas) [inline libraries/src/Plugin/PluginHelper.php]: You execute one `UPDATE` query, which is atomic. IMO you do not need to call `$db->lockTable()`. Checking the result of `$db->getAffectedRows()` should be enough.

Comment (alikon) [reply libraries/src/Plugin/PluginHelper.php]: good point, maybe i was too scared about concurrency issues.... :face_with_head_bandage: 
we can changed it for sure,  if the whole thing gain some kind of green light

**Comment by PhilETaylor** [comment]
Only just seen this. These are just some thoughts. 

Literally every other major PHP app has crons

If we are going to have a cron based system then it should be able to accept `* * * * *` based notation as well as or in place of "beginners select 1min from a dropdown".  There are lots of great libs that can parse the notation 

I wrote the original cron processes for Comic Relief, they were based on the magento cron scheduler back then, but there are lots of other great examples. 

The concept should be that you can set a server/web cron for every min of every day and then let the scheduler within the PHP app decide what jobs need to run. They you dont need to keep changing your cron settings when you add a new extension/job.

The Joomla API could expose an API to add new jobs (e.g through an extension or plugin job like you have) but it should be the scheduler and not the job that decides if it should run. E.g I should not have to load 100 jobs and ask each one if they need to run this time. The scheduler should know the frequency of work for each job and only load the jobs it needs to on each pass. 

As for concurrency issues there are two chains of thought, 1) Ignore it, its not the crons fault if a job takes so long its still running when the next job is due, 2) use something like [symfony/lock](https://symfony.com/doc/current/components/lock.html) in semaphore or the new mysql mode - or even hook that to memcache/redis if Joomla has that configured.  

As for what "triggers" the scheduler every min, well like Wordpress this could be an after render process activated by default on all new joomla installs - and then for advanced users, they could disable this and trigger using webcron services, or webhosts cli crontab. ALL should be possible. Yes the limitations on the after render process are known, but if your jobs are so time critical that you cannot wait for the next page load (bot or visitor) then you should use one of the other methods. Most cron jobs are NOT time critical to that degree (eg, log rotation, Garbage collection)

There should also be a cron overview page so you can see all cron jobs in one place, rather than just filtering on "job plugins" 

This is the kind of interface Joomla could provide (magento example)
available_tasks

Here is the real life live example from the myJoomla.com service (truncated) 

screenshot 2018-09-29 at 11 55 22

File: administrator/language/en-GB/en-GB.plg_job_logrotation.ini [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+; Note : All ini files need to be saved as UTF-8
+
+PLG_JOB_LOGROTATION="Job - Log Rotation"
+PLG_JOB_LOGROTATION_XML_DESCRIPTION="This plugin periodically rotates system log files."
+PLG_JOB_LOGROTATION_LOGSTOKEEP_DESC="The maximum number of old logs to keep."
+PLG_JOB_LOGROTATION_LOGSTOKEEP_LABEL="Maximum Logs"
+PLG_JOB_LOGROTATION_CACHETIMEOUT_LABEL="Frequency (in unit)"
+PLG_JOB_LOGROTATION_CACHETIMEOUT_DESC="How often should the job run."
+PLG_JOB_LOGROTATION_UNIT_LABEL="Unit of time"
+PLG_JOB_LOGROTATION_UNIT_DESC="How often should the job run."
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.plg_job_logrotation.ini]: This isnt a correct description for the selection of the unit

File: plugins/job/logrotation/logrotation.xml [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+				/>
+
+				<field
+					name="unit"
+					type="list"
+					label="PLG_JOB_LOGROTATION_UNIT_LABEL"
+					description="PLG_JOB_LOGROTATION_UNIT_DESC"
+					default="86400"
+					filter="integer"
+				>
```
Comment (brianteeman) [inline plugins/job/logrotation/logrotation.xml:62]: indent one tab

Comment (alikon) [reply plugins/job/logrotation/logrotation.xml:62]: fixed

File: administrator/language/en-GB/en-GB.plg_job_logrotation.ini (hunk 10 above)
Comment (alikon) [reply administrator/language/en-GB/en-GB.plg_job_logrotation.ini]: changed

File: libraries/src/Plugin/PluginHelper.php (hunk 9 above)
Comment (alikon) [reply libraries/src/Plugin/PluginHelper.php]: let's go with the "atomic"

File: plugins/job/logrotation/logrotation.xml (hunk 11 above)
Comment (Quy) [reply plugins/job/logrotation/logrotation.xml:62]: Indent lines 62-65.

**Comment by tecpromotion** [comment]
> @roland-d should be noticed that drupal has a "poor mans cron" see https://www.drupal.org/docs/8/cron-automated-tasks/cron-automated-tasks-overview

Shopware has one, TYPO3 of course.
In both examples, it is more about setting up the cronjobs, ie the tasks and how often they should run.
It must then continue to run a real cron job on the host, which calls the scheduler. For example. every 5 minutes, the scheduler is called up and then looks if one of the jobs has to be processed and leaves it running. In Shopware, the answer will be displayed partly in the settings of the respective task with values, or a true false. It can even be set to be notified by false via email or even the task is set to inactive.

**Comment by ghost** [comment]
As new Feature this should be rebased on J4.

**Comment by HLeithner** [comment]
@alikon I would be very happy to see this feature in J4. But wouldn't be merged in J3. 

I close this here, please reference if you have a PR for J4.

**Comment by brianteeman** [comment]
@alikon it would be really great if you were able to port this to J4

**Pull request closed without merging**